File: common_subsets.rb
require 'byebug';

def common_subsets(arr1, arr2) 
    subs1 = subsets(arr1)
    subs2 = subsets(arr2)

    common = []
    subs2.each {|ele| common << ele if subs1.include?(ele)}
    common
end

def subsets(arr)
    return [[]] if arr.empty? 
    debugger
    first = arr[0]
    subset = subsets(arr.drop(1))
    new_subs = subset.map {|ele| ele + [first]}
    new_subs + subset
end

p common_subsets([1,2,3], [2,3])

File: pairboarding/common_subsets.js
function commonSubsets(arr1, arr2) {
    let intersection = fastIntersection(arr1, arr2);

    return subsets(intersection);
}

function subsets(arr) {
    if(arr.length === 0) return [[]];

    let first = arr.shift();
    let subs = subsets(arr);

    let rem = subs.map((ele) => ele.concat([first]));

    return subs.concat(rem);
}

function fastIntersection(arr1, arr2) {
    let intersection = [];
    let set = new Set(arr1);

    for (let i = 0; i < arr2.length; i++) {
        if (set.has(arr2[i])) intersection.push(arr2[i]);
    }

    return intersection;
}

console.log(commonSubsets([1,2,3], [1,2]))